### ISSUE TITLE
DirectoryLister in twisted.web.static does not call base constructor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3631]
|<img alt="eulores's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| eulores reported|
|-|-|
|Trac ID|trac#3631|
|Type|defect|
|Created|2009-01-23 17:44:40Z|
|Branch|https://github.com/twisted/twisted/tree/directory-lister-children-3631|

`DirectoryLister` in `twisted.web.static` does not call base constructor, including something along the lines of
```
resource.Resource.__init__(self)
```
in the constructor of `DirectoryLister` is enough.

This popped up while studying `static.File()` behavior on URLS that have embedded double slashes, like `http://example.com/my/path//has/double/slashes`

<details><summary>Searchable metadata</summary>

```
trac-id__3631 3631
type__defect defect
reporter__eulores eulores
priority__normal normal
milestone__ 
branch__branches_directory_lister_children_3631 branches/directory-lister-children-3631
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__easy easy
time__1232732680000000 1232732680000000
changetime__1234652692000000 1234652692000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="eulores's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|eulores commented|
|-|-|

The traceback I got is
`<type 'exceptions.AttributeError'>: DirectoryLister instance has no attribute 'children'`

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [26194]) Branching to 'directory-lister-children-3631'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [26195]) test for File.getChild for the empty resource, test for DirectoryLister.getChild; fix for DirectoryLister.getChild

refs [#3631](https://github.com/twisted/twisted/issues/3631)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/directory-lister-children-3631)

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Cool, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [26221]) Merge directory-lister-children-3631

Author: exarkun
Reviewer: therve
Fixes: [#3631](https://github.com/twisted/twisted/issues/3631)

Fix `DirectoryLister` so that it allows `Resource` (its base class) to properly initialize
itself.  This fixes an `AttributeError` raised when looking up a child of `DirectoryLister`.

Additionally, some of the tests have been changed to use `resource.getChildForRequest` rather
than calling `getChild` directly, as the former is a more realistic approximation of what
happens in a real application.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
